### PR TITLE
[perf] nn.SyncBatchNorm: use autograd function to save memory

### PR DESCRIPTION
**Pull request by msbaines**
## What does this PR do?
Fixes # (issue).

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by min-xu-ai** [comment]
why does this save memory?

**Comment by msbaines** [comment]
> why does this save memory?

By using an autograd function you have control over which intermediate results are stored during forward. Without an autograd function, all intermediate results are captured by the autograd graph.

Here we use an autograd graph to store some intermediate results like mean and invstd which consume little memory relative to BatchNorm input. The downside is that we need to do some re-computation in order to calculate gradients. The memory saving is quite large.

@endernewton reported his model OOMing when replacing torch's SyncBatchNorm with this one. The memory overhead of this implementation should now be on part with torch's.

File: tests/experimental/nn/test_sync_batchnorm.py [hunk 1]
```diff
@@ -36,40 +36,48 @@ def wrapper(*args, **kwargs):
...
+    fs_y = fs_bn(fs_x)
     fs_y.backward(yh)
     assert torch.allclose(torch_y, fs_y), f"{torch_y} != {fs_y}"
     assert torch.allclose(torch_bn.running_mean, fs_bn.running_mean), f"{torch_bn.running_mean} != {fs_bn.running_mean}"
     assert torch.allclose(torch_bn.running_var, fs_bn.running_var), f"{torch_bn.running_var} != {fs_bn.running_var}"
-    assert torch.allclose(torch_bn.weight, fs_bn.weight), f"{torch_bn.weight.grad} != {fs_bn.weight.grad}"
-    assert torch.allclose(torch_bn.bias, fs_bn.bias), f"{torch_bn.bias.grad} != {fs_bn.bias.grad}"
-    # TODO(msb) currently disabled due to PyTorch bug: https://github.com/pytorch/pytorch/issues/57796
-    # assert torch.allclose(torch_bn.weight.grad, fs_bn.weight.grad), f"{torch_bn.weight.grad} != {fs_bn.weight.grad}"
+    assert torch.allclose(torch_bn.weight, fs_bn.weight), f"{torch_bn.weight} != {fs_bn.weight}"
```
Comment (min-xu-ai) [inline tests/experimental/nn/test_sync_batchnorm.py:51]: btw, torch.testing.assert_allclose() is a much better method on asserting tensors.

Comment (msbaines) [reply tests/experimental/nn/test_sync_batchnorm.py:51]: Wow! torch.testing.assert_allclose() is amazing. Thanks for introducing me to it!

File: tests/experimental/nn/test_sync_batchnorm.py [hunk 2]
```diff
@@ -36,40 +36,48 @@ def wrapper(*args, **kwargs):
 
 
 def check_parity(torch_bn, fs_bn, x):
-    yh = torch.ones_like(x)
-    torch_y = torch_bn(x)
-    fs_y = fs_bn(x)
+    yh = torch.randn_like(x)
+    torch_x = x.detach()
```
Comment (anj-s) [inline tests/experimental/nn/test_sync_batchnorm.py:40]: why do you need to detach input here?

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 3]
```diff
@@ -62,6 +40,87 @@ def _forward(
...
+                grad_input = invstd * grad_output
+                grad_mean = -torch.sum(grad_input, dim=dim, keepdim=True)
+                grad_invstd = grad_common
+            else:
+                grad_input = (invstd * weight.reshape_as(mean)) * grad_output
+                grad_mean = -torch.sum(grad_input, dim=dim, keepdim=True)
+                grad_invstd = grad_common * weight.reshape_as(mean)
+            grad_var = -0.5 * invstd.pow(3) * grad_invstd
+            grad_mean += -2 * mean * grad_var
+            grad_meansqr = grad_var
```
Comment (anj-s) [inline fairscale/experimental/nn/sync_batchnorm.py:106]: why not use grad_var directly instead of reassigning to grad_meansqr?

**Comment by msbaines** [comment]
ping @min-xu-ai @anj-s

File: fairscale/experimental/nn/sync_batchnorm.py (hunk 3 above)
Comment (msbaines) [reply fairscale/experimental/nn/sync_batchnorm.py:106]: For readability. I wrote this logic with the chain rule in mind. So first I calculate `grad_var` and then I use the chain rule to calculate `grad_meansqr`. If I directly assigned `grad_meansqr` then the expression for `grad_mean` would look odd.

File: tests/experimental/nn/test_sync_batchnorm.py (hunk 2 above)
Comment (msbaines) [reply tests/experimental/nn/test_sync_batchnorm.py:40]: Because I want `torch_x` and `fs_x` to share the same data but have different gradients. I guess I could have used `clone()` instead and just created multiple copies of the `x` tensor.

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 4]
```diff
@@ -9,51 +9,29 @@
 import torch.distributed as dist
 from torch.distributed import ProcessGroup
```
Comment (min-xu-ai) [inline fairscale/experimental/nn/sync_batchnorm.py:10]:
```suggestion
from torch.distributed import ProcessGroup

from fairscale.utils.testing import torch_version
```

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 5]
```diff
@@ -62,6 +40,87 @@ def _forward(
     _forward = torch.jit.script(_forward)  # type: ignore
 
 
+class _SyncBatchNormFunction(torch.autograd.Function):
```
Comment (min-xu-ai) [inline fairscale/experimental/nn/sync_batchnorm.py:43]: on line 39, I think you can use torch_version() function

Comment (min-xu-ai) [reply fairscale/experimental/nn/sync_batchnorm.py:43]: we can move that function our to testing.py as well since we do use it in non testing code.

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 6]
```diff
@@ -62,6 +40,87 @@ def _forward(
...
+    # type: ignore
+    def forward(
+        ctx, input, weight, bias, affine, track_running_stats, running_mean, running_var, eps, momentum, process_group
+    ):
+        dim = [d for d in range(input.ndim) if d != 1]
+        count = torch.full((1,), input.numel() // input.size(1), device=input.device, dtype=input.dtype)
+        total_count = count.clone()
+        handle = dist.all_reduce(total_count, group=process_group, async_op=True)
+        mean = torch.mean(input, dim=dim, keepdim=True)
+        meansqr = torch.mean(input * input, dim=dim, keepdim=True)
```
Comment (min-xu-ai) [inline fairscale/experimental/nn/sync_batchnorm.py:54]: it seems that input.pow(2) is faster?

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 7]
```diff
@@ -62,6 +40,87 @@ def _forward(
...
+class _SyncBatchNormFunction(torch.autograd.Function):
+    @staticmethod
+    # type: ignore
+    def forward(
+        ctx, input, weight, bias, affine, track_running_stats, running_mean, running_var, eps, momentum, process_group
+    ):
+        dim = [d for d in range(input.ndim) if d != 1]
+        count = torch.full((1,), input.numel() // input.size(1), device=input.device, dtype=input.dtype)
+        total_count = count.clone()
+        handle = dist.all_reduce(total_count, group=process_group, async_op=True)
```
Comment (min-xu-ai) [inline fairscale/experimental/nn/sync_batchnorm.py:52]:
```suggestion
        all_reduce_handle = dist.all_reduce(total_count, group=process_group, async_op=True)
```

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 8]
```diff
@@ -62,6 +40,87 @@ def _forward(
...
+        mean = torch.mean(input, dim=dim, keepdim=True)
+        meansqr = torch.mean(input * input, dim=dim, keepdim=True)
+        vec = torch.cat([mean, meansqr])
+        handle.wait()
```
Comment (min-xu-ai) [inline fairscale/experimental/nn/sync_batchnorm.py:56]:
```suggestion
        all_reduce_handle.wait()
```

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 9]
```diff
@@ -62,6 +40,87 @@ def _forward(
...
+        vec = vec * (count / total_count)
+        dist.all_reduce(vec, group=process_group)
+        mean, meansqr = vec.chunk(2)
+        var = meansqr - mean * mean
```
Comment (min-xu-ai) [inline fairscale/experimental/nn/sync_batchnorm.py:60]: mean.pow(2)?

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 10]
```diff
@@ -62,6 +40,87 @@ def _forward(
...
+        input, weight, bias, mean, invstd, total_count = ctx.saved_tensors
+        process_group = ctx.process_group
+        dim = [d for d in range(input.ndim) if d != 1]
+        if ctx.needs_input_grad[0] or ctx.needs_input_grad[1]:
+            grad_common = torch.sum(
+                (input - mean) * grad_output, dim=dim, keepdim=True
+            )  # common to grad_weight and grad_invstd
+
+        if ctx.needs_input_grad[0]:
+            if weight is None:
```
Comment (min-xu-ai) [inline fairscale/experimental/nn/sync_batchnorm.py:96]: add a comment on why weight can be none?

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 11]
```diff
@@ -62,6 +40,87 @@ def _forward(
...
+    @staticmethod
+    # type: ignore
+    def backward(ctx, grad_output):
+        grad_input = None
+        grad_weight = None
+        grad_bias = None
+        input, weight, bias, mean, invstd, total_count = ctx.saved_tensors
+        process_group = ctx.process_group
+        dim = [d for d in range(input.ndim) if d != 1]
+        if ctx.needs_input_grad[0] or ctx.needs_input_grad[1]:
```
Comment (min-xu-ai) [inline fairscale/experimental/nn/sync_batchnorm.py:90]: since this needs_input_grad list seems to be maintain somewhere else, it is better to add a comment and document their meaning and whether one of them must be true or they can be both true or both false?

Comment (min-xu-ai) [reply fairscale/experimental/nn/sync_batchnorm.py:90]: will be useful for people making changes here in the future.

**Review by min-xu-ai — APPROVED** [review]
very good! some minor and nonblocking comments only.

File: tests/experimental/nn/test_sync_batchnorm.py (hunk 2 above)
Comment (anj-s) [reply tests/experimental/nn/test_sync_batchnorm.py:40]: I thought detach removes the variable from the graph but I guess you then specify requires_grad which makes it a part of the graph again. Either way is fine with me. I guess I was not familiar with the usage.

**Review by anj-s — APPROVED** [review]

**Comment by msbaines** [comment]
Added a memory usage test. I get:

```
fairscale.experimental.nn.SyncBatchNorm: 20996608 
torch.nn.SyncBatchNorm: 21021696
```

File: fairscale/experimental/nn/sync_batchnorm.py (hunk 11 above)
Comment (msbaines) [reply fairscale/experimental/nn/sync_batchnorm.py:90]: Good idea. needs_input_grad is maintained by autograd function base class.

File: fairscale/experimental/nn/sync_batchnorm.py (hunk 6 above)
Comment (msbaines) [reply fairscale/experimental/nn/sync_batchnorm.py:54]: Cool. Do you have a reference?

Comment (min-xu-ai) [reply fairscale/experimental/nn/sync_batchnorm.py:54]: I don’t have one. I just did some test in ipython which seems faster but that could be noise.
